File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

MOVES = [(-1, 0), (1, 0), (0, -1), (0, 1)]

def in_bound(nx, ny):
    return 0 <= nx < 5 and 0 <= ny < 5

def check_manhattan_dist(x1, y1, x2, y2):
    return True if abs(x1 - x2) + abs(y1 - y2) <= 2 else False

def bfs(i, j, place):
    queue = deque([(i, j)])
    
    visited = [[False for _ in range(5)] for _ in range(5)]
    visited[i][j] = True
                
    while queue:
        x, y = queue.popleft()
        
        for dx, dy in MOVES:
            nx, ny = x + dx, y + dy
            
            if in_bound(nx, ny) and not visited[nx][ny] and check_manhattan_dist(i, j, nx, ny):
                # P가 있으면 바로 종료
                if place[nx][ny] == 'P':
                    return False
                # 파티션이 있으면 그 방향으로는 탐색 종료
                elif place[nx][ny] == 'X':
                    continue
                else:
                    queue.append((nx, ny))
                    visited[nx][ny] = True
                
    return True

def solution(places):
    answer = []
    
    for place in places:    
        is_ok = True
                
        for i in range(5):
            for j in range(5):
                if place[i][j] == 'P':
                    if not bfs(i, j, place):
                        is_ok = False
                        break
                
            if not is_ok:
                break
        
        answer.append(1 if is_ok else 0)
                
    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.kt
import kotlin.math.abs

class Solution {
    private val MOVES = listOf(-1 to 0, 1 to 0, 0 to -1, 0 to 1)
    
    data class Point(val x: Int, val y: Int)
    
    fun inBound(nx: Int, ny: Int): Boolean {
        return (0 <= nx && nx < 5) && (0 <= ny && ny < 5)
    }
    
    fun checkManhattan(x1: Int, y1: Int, x2: Int, y2: Int): Boolean {
        if (abs(x1 - x2) + abs(y1 - y2) <= 2) {
            return true
        } else {
            return false
        }
    }
    
    fun bfs(i: Int, j: Int, place: Array<String>): Boolean {
        val queue = ArrayDeque<Point>()
        queue.add(Point(i, j))
        
        val visited = Array(5) {BooleanArray(5)}
        visited[i][j] = true
        
        while (queue.isNotEmpty()) {
            val (x, y) = queue.removeFirst()
            
            for ((dx, dy) in MOVES) {
                val nx = x + dx
                val ny = y + dy
                
                if (inBound(nx, ny) && !visited[nx][ny] && checkManhattan(i, j, nx, ny)) {
                    if (place[nx][ny] == 'P') {
                        return false
                    } else if (place[nx][ny] == 'X') {
                        continue
                    } else {
                        queue.add(Point(nx, ny))
                        visited[nx][ny] = true
                    }
                }
            }
        }
        
        return true
    }
    
    fun solution(places: Array<Array<String>>): IntArray {
        var answer = mutableListOf<Int>()
        
        for (place in places) {
            var isOk = true
            
            outer@ for (i in 0 until 5) {
                for (j in 0 until 5) {
                    if (place[i][j] == 'P') {
                        if (!bfs(i, j, place)) {
                            isOk = false
                            break@outer
                        }
                    }
                }
            }
            
            answer.add(if (isOk) 1 else 0)
        }
        
        return answer.toIntArray()
    }
}